File: solutions/0005_longest-palindromic-substring.py
# Problem: https://leetcode.com/problems/longest-palindromic-substring
# Runtime: 649 ms

class Solution:
    def longestPalindrome(self, s: str) -> str:
        start = 0
        end = 0
        for i in range(len(s)):
            curr = s[i]
            d = 1
            # odd
            while i - d >= 0 and i + d < len(s):
                c1 = s[i - d]
                c2 = s[i + d]
                if c1 != c2:
                    break
                if 2 * d + 1 > end - start:
                    start = i - d
                    end = i + d
                d += 1
            # even
            d = 0
            while i - d >= 0 and i + d + 1 < len(s):
                c1 = s[i - d]
                c2 = s[i + d + 1]
                if c1 != c2:
                    break
                if 2 * d + 2 > end - start:
                    start = i - d
                    end = i + d + 1
                d += 1
        return s[start : end + 1]

File: solutions/0005_longest-palindromic-substring.rb
# Problem: https://leetcode.com/problems/longest-palindromic-substring
# Runtime: 586 ms

# @param {String} s
# @return {String}
def longest_palindrome(s)
    return "" if s.length == 0
    longest = 1
    indices = [0, 0]
    # first, check all possible centers of odd length palindromes 
    1.upto(s.length - 1) do |ctr|
        l = ctr - 1
        r = ctr + 1
        until l < 0 or r >= s.length()
            break if s[l] != s[r]
            if r - l + 1 > longest
                longest = r - l + 1
                indices = [l, r]
            end
            l -= 1
            r += 1
        end
    end
    # next, check odd length palindromes
    0.upto(s.length - 2) do |ctr| # center is after s[ctr]
        l = ctr + 1
        r = ctr
        until l < 0 or r >= s.length()
            break if s[l] != s[r]
            if r - l + 1 > longest
                longest = r - l + 1
                indices = [l, r]
            end
            l -= 1
            r += 1
        end
    end
    return s[indices[0]..indices[1]]
end